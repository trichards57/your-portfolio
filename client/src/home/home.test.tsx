import React, { forwardRef as mockForwardRef } from "react";
import { fireEvent, render, waitFor } from "@testing-library/react";
import ReactDOM from "react-dom";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { useAuth0 } from "@auth0/auth0-react";
import { useHistory } from "react-router-dom";
import { HomeBase } from ".";

jest.mock("@auth0/auth0-react");
jest.mock("react-router-dom", () => ({
  Link: mockForwardRef((p: any, ref) => <div ref={ref} {...p} />),
  useHistory: jest.fn(),
}));

const testToken = "abcdefg";

const recentEndPoint = "/api/RecentShifts";
const allEndPoint = "/api/GetAllShifts";

const testRecentData = [
  {
    date: "2020-01-01",
    duration: 3,
    event: "Test Shift 1",
    id: "abcdef",
    loggedCalls: 4,
    role: "EAC",
    crewMate: "Test Crewmate 1",
    location: "Test Location 1",
  },
  {
    date: "2020-02-02",
    duration: 5,
    event: "Test Shift 2",
    id: "abcfef",
    loggedCalls: 4,
    role: "AFA",
    location: "Test Location 2",
  },
  {
    date: "2020-01-03",
    duration: 8,
    event: "Test Shift 3",
    id: "abcgef",
    loggedCalls: 4,
    role: "CRU",
    crewMate: "Test Crewmate 3",
  },
];

const testAllData = Array.from(Array(20).keys()).map((i) => ({
  date: `2020-01-${(i + 1).toString().padStart(2, "0")}`,
  duration: i / 2,
  event: `Test Shift ${i}`,
  id: `abcd${i.toString().padStart(2, "0")}`,
  loggedCalls: i % 5,
  role: "EAC",
  crewMate: `Test Crewmate ${i}`,
  location: `Test Location ${i}`,
}));

const server = setupServer(
  rest.get(recentEndPoint, (req, res, ctx) => {
    if (req.headers.get("authorization") === `Bearer ${testToken}`) {
      return res(ctx.json(testRecentData));
    }

    return res(ctx.status(401));
  }),
  rest.get(allEndPoint, (req, res, ctx) => {
    if (req.headers.get("authorization") === `Bearer ${testToken}`) {
      const pageString = req.url.searchParams.get("page");
      const countString = req.url.searchParams.get("count");

      if (!pageString || !countString) return res(ctx.status(400));

      const page = parseInt(pageString, 10);
      const count = parseInt(countString, 10);

      const data = testAllData.slice(page * count, (page + 1) * count);

      return res(
        ctx.set("x-total-count", testAllData.length.toString()),
        ctx.json(data)
      );
    }

    return res(ctx.status(401));
  })
);

beforeAll(() => server.listen());

beforeEach(() =>
  (useAuth0 as jest.Mock).mockReturnValue({
    getAccessTokenSilently: jest
      .fn()
      .mockReturnValue(Promise.resolve(testToken)),
  })
);

afterEach(() => server.resetHandlers());

afterAll(() => server.close());

it("renders without crashing", () => {
  const div = document.createElement("div");

  ReactDOM.render(<HomeBase />, div);

  ReactDOM.unmountComponentAtNode(div);
});

it("renders all without crashing", () => {
  const div = document.createElement("div");

  ReactDOM.render(<HomeBase all />, div);

  ReactDOM.unmountComponentAtNode(div);
});

it("renders loading correctly", () => {
  const res = render(<HomeBase />);

  expect(res.asFragment()).toMatchSnapshot();
});

it("renders loading all correctly", () => {
  const res = render(<HomeBase all />);

  expect(res.asFragment()).toMatchSnapshot();
});

it("renders successful load correctly", async () => {
  const res = render(<HomeBase />);

  await res.findByText("Test Shift 1");

  expect(res.asFragment()).toMatchSnapshot();
});

it("renders successful load all correctly", async () => {
  const res = render(<HomeBase all />);

  await res.findByText("Test Shift 1");

  expect(res.asFragment()).toMatchSnapshot();
});

it("redirects to the home page if not authorised", async () => {
  (useAuth0 as jest.Mock).mockReturnValue({
    getAccessTokenSilently: jest
      .fn()
      .mockReturnValue(Promise.resolve("BadToken")),
  });

  const testPush = jest.fn();

  (useHistory as jest.Mock).mockReturnValue({
    push: testPush,
  });

  render(<HomeBase />);

  await waitFor(() => expect(testPush).toBeCalledWith("/"));
});

it("redirects to the home page if all not authorised", async () => {
  (useAuth0 as jest.Mock).mockReturnValue({
    getAccessTokenSilently: jest
      .fn()
      .mockReturnValue(Promise.resolve("BadToken")),
  });

  const testPush = jest.fn();

  (useHistory as jest.Mock).mockReturnValue({
    push: testPush,
  });

  render(<HomeBase all />);

  await waitFor(() => expect(testPush).toBeCalledWith("/"));
});

it("displays an error if the request fails", async () => {
  server.use(rest.get(recentEndPoint, (_, res, ctx) => res(ctx.status(500))));

  const res = render(<HomeBase />);

  await res.findByText("There was a problem speaking to the server.", {
    exact: false,
  });
});

it("displays an error if request alls fails", async () => {
  server.use(rest.get(recentEndPoint, (_, res, ctx) => res(ctx.status(500))));

  const res = render(<HomeBase />);

  await res.findByText("There was a problem speaking to the server.", {
    exact: false,
  });
});

it("loads a new page if pagination clicked", async () => {
  const res = render(<HomeBase all />);

  await res.findByText("Test Shift 1");

  fireEvent.click(
    res.getByRole("button", {
      name: "Go to next page",
    })
  );

  await res.findByText("Test Shift 7");

  expect(res.asFragment()).toMatchSnapshot();
});

it("loads all if switches", async () => {
  const res = render(<HomeBase />);

  await res.findByText("Test Shift 1");

  res.rerender(<HomeBase all />);

  await res.findByText("Test Shift 1");

  expect(res.asFragment()).toMatchSnapshot();
});
