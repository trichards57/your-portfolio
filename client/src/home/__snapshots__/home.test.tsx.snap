// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders successful load correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-11"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-13 mui-fixed MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          aria-label="open drawer"
          class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-14 MuiIconButton-colorInherit MuiIconButton-edgeStart"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap"
        >
          Your Portfolio
        </h6>
      </div>
    </header>
    <nav
      aria-label="mailbox folders"
      class="makeStyles-drawer-12"
    />
    <main
      class="makeStyles-content-17"
    >
      <div
        class="makeStyles-toolbar-15"
      />
      <h2
        class="MuiTypography-root MuiTypography-h5"
      >
        Recent Shifts
      </h2>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-6 MuiGrid-grid-lg-3"
        >
          <div
            class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiCardContent-root"
            >
              <p
                class="MuiTypography-root makeStyles-date-21 makeStyles-trimText-22 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
              >
                1 January 2020
              </p>
              <h2
                class="MuiTypography-root makeStyles-trimText-22 MuiTypography-h5"
              >
                Test Shift 1
              </h2>
              <p
                class="MuiTypography-root makeStyles-location-23 makeStyles-trimText-22 MuiTypography-body1 MuiTypography-colorTextSecondary"
              >
                Test Location 1
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                Role: EAC
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                Logged Calls: 4
              </p>
            </div>
            <div
              class="MuiCardActions-root MuiCardActions-spacing"
            >
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
                role="button"
                tabindex="0"
                to="/editShift/abcdef"
              >
                <span
                  class="MuiButton-label"
                >
                  Edit
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
                role="button"
                tabindex="0"
                to="/addJob?shift=abcdef"
              >
                <span
                  class="MuiButton-label"
                >
                  Add Job
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-6 MuiGrid-grid-lg-3"
        >
          <div
            class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiCardContent-root"
            >
              <p
                class="MuiTypography-root makeStyles-date-21 makeStyles-trimText-22 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
              >
                3 January 2020
              </p>
              <h2
                class="MuiTypography-root makeStyles-trimText-22 MuiTypography-h5"
              >
                Test Shift 3
              </h2>
              <p
                class="MuiTypography-root makeStyles-location-23 makeStyles-trimText-22 MuiTypography-body1 MuiTypography-colorTextSecondary"
              />
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                Role: CRU
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                Logged Calls: 4
              </p>
            </div>
            <div
              class="MuiCardActions-root MuiCardActions-spacing"
            >
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
                role="button"
                tabindex="0"
                to="/editShift/abcgef"
              >
                <span
                  class="MuiButton-label"
                >
                  Edit
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
                role="button"
                tabindex="0"
                to="/addJob?shift=abcgef"
              >
                <span
                  class="MuiButton-label"
                >
                  Add Job
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-6 MuiGrid-grid-lg-3"
        >
          <div
            class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiCardContent-root"
            >
              <p
                class="MuiTypography-root makeStyles-date-21 makeStyles-trimText-22 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
              >
                2 February 2020
              </p>
              <h2
                class="MuiTypography-root makeStyles-trimText-22 MuiTypography-h5"
              >
                Test Shift 2
              </h2>
              <p
                class="MuiTypography-root makeStyles-location-23 makeStyles-trimText-22 MuiTypography-body1 MuiTypography-colorTextSecondary"
              >
                Test Location 2
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                Role: AFA
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2"
              >
                Logged Calls: 4
              </p>
            </div>
            <div
              class="MuiCardActions-root MuiCardActions-spacing"
            >
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
                role="button"
                tabindex="0"
                to="/editShift/abcfef"
              >
                <span
                  class="MuiButton-label"
                >
                  Edit
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
                role="button"
                tabindex="0"
                to="/addJob?shift=abcfef"
              >
                <span
                  class="MuiButton-label"
                >
                  Add Job
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
